src/server/storage_engine/block_engine/allocator.rs

```diff
@@ -2,54 +2,311 @@
 
-//#define BLKGETSIZE _IO(0x12,96)	/* return device size /512 (long *arg) */
+//#define BLKGETSIZE _IO(0x12,96)   /* return device size /512 (long *arg) */
 const BLOCKGETSIZE: u64 = 0x1260;
 
 pub const CHUNK: u64 = 512 * 8;
 const SECTOR: u64 = 512;
 
 pub(crate) trait Allocator {
     fn new(path: &str) -> Self;
-    fn allocator_space(&self, lenth: u64) -> u64;
+    fn allocator_space(&self, lenth: u64) -> Vec<AllocatorEntry>;
+    fn release_space(&self, lenth: u64, begin: u64) -> bool;
 }
 
 /*
- * This Allocator use for memory.
+ *This Allocator use for memory.
  */
 #[allow(unused)]
-pub(crate) struct BitmapAllocator {
-    block_space: Arc<Mutex<u64>>,
-    total_aspce: u64,
+pub(crate) struct BlockAllocator {
+    block_group_list: ArrayQueue<BlockGroup>,
+}
+
+struct BlockGroup {
+    block_space_point: u64,
+    block_space: u64,
+    slice_space_manager: Vec<HashSet<u64>>,
+    slice_space: u64,
 }
 
-impl Allocator for BitmapAllocator {
+impl Allocator for BlockAllocator {
     fn new(path: &str) -> Self {
-        let blockdevice = BlockDevice::new(path).unwrap();
+        let block_device = BlockDevice::new(path).unwrap();
+        let num_cpus = num_cpus::get();
+        let block_group_list = ArrayQueue::new(num_cpus);
+        for index in 1..num_cpus + 1 {
+            let block_allocator =
+                BlockGroup::new(block_device.chunk_num / num_cpus as u64, index as u64);
+            match block_group_list.push(block_allocator) {
+                Ok(_) => {}
+                Err(_) => {
+                    todo!()
+                }
+            };
+        }
+        Self { block_group_list }
+    }
+
+    fn allocator_space(&self, lenth: u64) -> Vec<AllocatorEntry> {
+        let mut block_group_count = 0;
+        loop {
+            let block_group = self.block_group_list.pop();
+            match block_group {
+                Some(mut block_group) => {
+                    let result_vec = block_group.allocator_space(lenth);
+                    match self.block_group_list.push(block_group) {
+                        Ok(_) => {}
+                        Err(_) => {
+                            todo!()
+                        }
+                    };
+                    if block_group_count >= num_cpus::get() {
+                        return Vec::new();
+                    }
+                    if result_vec.is_empty() {
+                        block_group_count += 1;
+                        continue;
+                    }
+                    return result_vec;
+                }
+                None => {
+                    continue;
+                }
+            }
+        }
+    }
+
+    fn release_space(&self, _lenth: u64, _begin: u64) -> bool {
+        // todo
+        true
+    }
+}
+
+impl BlockGroup {
+    fn new(size: u64, block_group_index: u64) -> Self {
+        let mut slice_space_vec = Vec::new();
+        for _ in 0..11 {
+            slice_space_vec.push(HashSet::new());
```

Comment:
Where do you initialize hash-sets in slice_space_vec?

Reply:
This data structure can be understood as a primary linked list, and each node contains other linked lists